### ISSUE TITLE
FEATURE REQUEST: Save/Load mechanism

### ISSUE THREAD

**Issue by ElectricErger** [daid/EmptyEpsilon#1043]
# Synopsis
"Homebrew GM"s tend to make a world that they want their people to explore, and this tends to come with RPG advancements (e.g. buying improvements). As of right now, if you want to resume a game later, either the server must be on, or you must remember all the relevant details (which is not exactly fun).

## Additional details
We can get pretty far with a Lua script that generates the world, but after you've played in the world, you have to update it with various things that have changed. While this is doable, a save/load function sounds like a better feature, and would encourage sharing worlds that less tech-savvy people can work with.

# Ask
A mechanism to save and load various game state data:
- Ship details (e.g. coolant, damage, number of torpedo tubes)
- Map pieces (e.g. stations (that haven't been blown-up), stray mines)

**Comment by ElectricErger** [comment]
I hope I did this correctly. I didn't see any other issues like it, and my group thought it would be cool.

**Comment by daid** [comment]
In the gm screen you can copy most of the gamestate as a script to your clipboard. It is not a full save&load, but does most of what you need i think.

**Comment by czenker** [comment]
Another Life Hack is to run EE inside [Virtual Box](https://www.virtualbox.org/) or similar. If you stop playing, pause the VM. You can shut down your PC and continue later by unpausing the VM. You can even create a snapshot of the VM to have multiple savegames of the same instance.

I have not tried it real life yet, but I don't see why it should not work. And this is the way I would do it.

Especially with scenarios, adding save and load capabilities multiplies the effort it takes to write a scenario. So I do not think this will ever happen.

**Comment by oznogon** [comment]
Not that it'll move the needle, but GM screen copy functions don't work on macOS builds.

**Comment by aBlueShadow** [comment]
@oznogon iirc, on linux you need xclip for it to work. Is it possible to use this (or a similar tool) on macos too?

**Comment by oznogon** [comment]
@aBlueShadow As the code in SP stands now, `Clipboard` functions don't do or return anything unless you're on Windows or Linux. Macs have `pbcopy` built in. Since it looks like the SP implementation is essentially just:

```
    FILE* pipe = popen("/usr/bin/xclip -i -selection clipboard -silent", "we");
    fwrite(value.c_str(), value.size(), 1, pipe);
    pclose(pipe);
```

It might be doable.

**Comment by daid** [comment]
Note that the linux implementation most likely only works for small pieces of text, not sure when it fails, but it will most likely fail after a certain size.

**Comment by Ownezx** [comment]
Would the ECS system make it more straight forward to save/load a game?

**Comment by daid** [comment]
> Would the ECS system make it more straight forward to save/load a game?

Yes, but also, no. Saving all the component states is a bit easier. But, the whole lua state is really hard to save, especially if you think about callbacks.

**Comment by Ownezx** [comment]
What if we offload the saving responsibility of the lua state to scripters with a "onSave" and "onLoad" call?

Another Idea was for people to have a lua master object scripters need to use if the want the save/load function, but that will make a mess of a lua object...
